content/courses/advanced-ada/chapters/aggregates.rst

```diff
@@ -8,11 +8,465 @@ Aggregates
+to components of the same type:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Aggregates.Rec_Aggregates_Others
+
+    with Custom_Records; use Custom_Records;
+
+    procedure Show_Record_Aggregates_Others is
+       Dummy : Integer_Float;
+    begin
+       --  OK: components selected by
+       --      others have Integer type.
+       Dummy := (Y | Z  => 256.0,
+                 others => 256);
+    end Show_Record_Aggregates_Others;
+
+In any case, writing :ada:`(others => <>)` is always accepted by the compiler
+because it simply selects the default value of each component, so the type of
+those values is unambiguous:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Aggregates.Rec_Aggregates_Others
+
+    with Custom_Records; use Custom_Records;
+
+    procedure Show_Record_Aggregates_Others is
+       Dummy : Integer_Float;
+    begin
+       Dummy := (others => <>);
+    end Show_Record_Aggregates_Others;
+
+This code compiles because :ada:`<>` uses the appropriate default value of each
+component.
+
+
+Record discriminants
+~~~~~~~~~~~~~~~~~~~~
+
+When a record type has discriminants, they must appear as components of an
+aggregate of that type. For example, consider this package:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Aggregates.Rec_Aggregate_Discriminant
+
+    package Points is
+
+       type Point_Dimension is (Dim_1, Dim_2, Dim_3);
+
+       type Point (D : Point_Dimension) is record
+          case D is
+          when Dim_1 =>
+             X1         : Integer;
+          when Dim_2 =>
+             X2, Y2     : Integer;
+          when Dim_3 =>
+             X3, Y3, Z3 : Integer;
+          end case;
+       end record;
+
+       procedure Display (P : Point);
+
+    end Points;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Points is
+
+       procedure Display (P : Point) is
+       begin
+          Put_Line (Point_Dimension'Image (P.D));
+
+          case P.D is
+          when Dim_1 =>
+             Put_Line ("  (X => "
+                       & Integer'Image (P.X1)
+                       & ")");
+          when Dim_2 =>
+             Put_Line ("  (X => "
+                       & Integer'Image (P.X2)
+                       & ",");
+             Put_Line ("   Y => "
+                       & Integer'Image (P.Y2)
+                      & ")");
+          when Dim_3 =>
+             Put_Line ("  (X => "
+                       & Integer'Image (P.X3)
+                       & ",");
+             Put_Line ("   Y => "
+                       & Integer'Image (P.Y3)
+                       & ",");
+             Put_Line ("   Z => "
+                       & Integer'Image (P.Z3)
+                       & ")");
+          end case;
+       end Display;
+
+    end Points;
+
+To write aggregates of the :ada:`Point` type, we have to specify the :ada:`D`
+discriminant as a component of the aggregate. Also, this component must be the
```

Comment:
After the sentence beginning with "To write aggregates" let me suggest another sentence (before the one starting with "Also, "):
The discriminant must be included in the aggregate |mdash| and must be static |mdash| because the compiler must be able to examine the aggregate to determine if it is both complete and consistent. All components must be accounted for one way or another, as usual. But in addition, references to those components whose existence depends on the discriminant's values must be consistent with the actual discriminant value used in the aggregate. For example, for type :ada:`Point`, an aggregate can only reference the :ada:`X3`, :ada:`Y3`, and :ada:`Z3` components when :ada:`Dim_3` is specified for the discriminant :ada:`D`, otherwise those three components don't exist in that aggregate.

Reply:
Very good description, thanks a lot! I've included it (almost) verbatim.